testdata/p4_16_samples/psa-portid-using-newtype2.p4

```diff
@@ -0,0 +1,293 @@
+    // No annotation needed here to get agent translation of
+    // egress_port, for same reason as in header packet_in_header_t
+    // above.
+    PortIdInHeader_t egress_port;
+}
+
+
+struct parsed_headers_t {
+    packet_in_header_t packet_in;
+    packet_out_header_t packet_out;
+    ethernet_t ethernet;
+    ipv4_t ipv4;
+}
+
+struct fabric_metadata_t {
+}
+
+action drop() {
+    mark_to_drop();
+}
+
+action nop() {
+}
+
+parser FabricParser(packet_in packet,
+                    out parsed_headers_t hdr,
+                    inout fabric_metadata_t meta,
+                    inout standard_metadata_t standard_metadata)
+{
+    state start {
+        transition accept;
+    }
+}
+
+control PacketIoIngress(inout parsed_headers_t hdr,
+                        inout fabric_metadata_t fabric_metadata,
+                        inout standard_metadata_t standard_metadata)
+{
+    apply {
+        if (hdr.packet_out.isValid()) {
+            // I know what I'm doing, I requested translation on
+            // packet_out.egress_port
+            standard_metadata.egress_spec =
+                PSA_PORT_HEADER_TO_ID(hdr.packet_out.egress_port);
+            hdr.packet_out.setInvalid();
+            exit;
+        }
+    }
+}
+
+control PacketIoEgress(inout parsed_headers_t hdr,
+                       inout fabric_metadata_t fabric_metadata,
+                       inout standard_metadata_t standard_metadata)
+{
+    apply {
+        // comparison of PortId_t to PortId_t is valid
+        if (standard_metadata.egress_port == PSA_CPU_PORT) {
+            hdr.packet_in.setValid();
+            // I know what I'm doing, I requested translation on
+            // packet_in.ingress_port
+            hdr.packet_in.ingress_port =
+                PSA_PORT_ID_TO_HEADER(standard_metadata.ingress_port);
+        }
+    }
+}
+
+control Filtering (inout parsed_headers_t hdr,
+                   inout fabric_metadata_t fabric_metadata,
+                   inout standard_metadata_t standard_metadata)
+{
+    table t {
+        key = {
+            // translated by P4Runtime agent because of annotation on PortId_t
+            standard_metadata.ingress_port: exact;
+        }
+        actions = { drop; nop; }
+    }
+
+    apply { t.apply(); }
+}
+
+control Forwarding (inout parsed_headers_t hdr,
+                    inout fabric_metadata_t fabric_metadata,
+                    inout standard_metadata_t standard_metadata)
+{
+    // next_port translated by P4Runtime agent because of annotation on PortId_t
+    action fwd(PortId_t next_port) {
+        standard_metadata.egress_spec = next_port;
+    }
+
+    table t {
+        key = {
+            hdr.ipv4.dstAddr : exact;
+        }
+        actions = { drop; fwd; }
+    }
+
+    apply {
+        t.apply();
+
+        // One can perform arithmetic operations on PortId_t values if
+        // you use enough casts to convert it to an unsigned int, do
+        // the arithetic ops, then cast back to PortId_t.
+
```

Comment:
In my view, this should not be required. We must be able to operate on the values of PortId_t as if they are bit<PORT_BITWIDTH>.

Reply:
Given the current p4c compiler as of 2018-Aug-19, it is not possible to do arithmetic on two values of type PortId_t directly.  That is something that could be enhanced later, if the compiler is changed to support arithmetic on two values with the same `type` name, and produces a value with that same `type` name type.